### PR TITLE
#19: Distinugish better between error causes

### LINKED ISSUE #19: Distinugish better between error causes
## Situation

Method "getHttp()" wraps HTTP status codes into an I/O error. This is not only conceptually wrong, but also prevents proper distinction between error causes.

## Acceptance Criteria

* Download and upload error are distinguished
* 404 is reported as "file not found"
* 403 is reported as "access denied"

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #19.

### CODE REVIEW

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java [hunk 1]
```diff
@@ -149,20 +146,20 @@ void testDownloadFileThrowsExceptionOnIllegalLocalPath(@TempDir final Path tempD
...
-            final String currentFirstLine = useA ? contentA : contentB;
+        final var scaleContentSizeBy = 10000000;
+        final var fileName = "replace_me.txt";
+        final var absolutePathInContainer = "/exa/data/bucketfs/bfsdefault/.dest/default/" + fileName;
+        final var contentA = "0123456789\n";
+        final var fileA = Files.writeString(tempDir.resolve("a.txt"), contentA.repeat(scaleContentSizeBy));
+        final var contentB = "abcdeABCDE\n";
+        final var fileB = Files.writeString(tempDir.resolve("b.txt"), contentB.repeat(scaleContentSizeBy));
+        final var bucket = getDefaultBucket();
+        for (var i = 1; i <= 10; ++i) {
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:157]: var and int are both 3 chars long. ;)

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
Left a few comments
Please, be careful with vars. In some places it's not really an improvement.

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 2]
```diff
@@ -118,45 +123,81 @@ protected String extendPathInBucketDownToFilename(final Path localPath, final St
...
+            Thread.currentThread().interrupt();
+            throw createDownloadInterruptedException(uri, DOWNLOAD);
         }
-        LOGGER.fine(() -> "Successfully downloaded file to \"" + localPath + "\"");
+    }
+
+    private BucketAccessException createDownloadIoException(final URI uri, final BucketOperation operation,
+            final IOException exception) {
+        return new BucketAccessException(messageBuilder("E-BFSJ-5")
+                .message("I/O error trying to {{operation}} \"{{URI}}\"", operation, uri).toString(), exception);
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:152]: Good to know.

File: src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java (hunk 1 above)
Comment (redcatbear) [reply src/test/java/com/exasol/bucketfs/SyncAwareBucketIT.java:157]: Did I mention I like strict patterns?

File: doc/changes/changes_2.0.0.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# bucketfs-java 2.0.0, released 2021-04-XX
+
+Code name:
+
+Breaking changes:
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/changes/changes_2.0.0.md:5]: Done.

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:152]: Fixed quoting in error messages.

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 4]
```diff
@@ -118,45 +123,81 @@ protected String extendPathInBucketDownToFilename(final Path localPath, final St
...
+        return response.body();
+    }
+
+    private HttpResponse<String> requestFileOnBucketAsString(final URI uri) throws BucketAccessException {
         try {
-            return httpGet(uri);
+            final var request = HttpRequest.newBuilder(uri) //
+                    .GET() //
+                    .header("Authorization", encodeBasicAuthForReading()) //
+                    .build();
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:175]: Fixed.

File: src/main/java/com/exasol/bucketfs/SyncAwareBucket.java [hunk 5]
```diff
@@ -26,103 +28,84 @@ protected SyncAwareBucket(final Builder<? extends Builder<?>> builder) {
...
     // Wait some time between uploads of the same file so we can distinguish the upload success logs for detecting
     // upload success.
     // [impl->dsn~bucketfs-object-overwrite-throttle~1]
-    private void delayRepeatedUploadToSamePath(final String extendedPathInBucket) throws InterruptedException {
+    private void delayRepeatedUploadToSamePath(final String extendedPathInBucket) throws BucketAccessException {
         if (this.uploadHistory.containsKey(extendedPathInBucket)) {
-            final Instant lastUploadAt = this.uploadHistory.get(extendedPathInBucket).with(ChronoField.NANO_OF_SECOND,
-                    0);
-            final Instant now = Instant.now();
+            final var lastUploadAt = this.uploadHistory.get(extendedPathInBucket).with(ChronoField.NANO_OF_SECOND, 0);
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/SyncAwareBucket.java:50]: Fixed.

File: src/test/java/com/exasol/bucketfs/BucketContentSyncIT.java [hunk 6]
```diff
@@ -42,15 +40,15 @@ private Bucket getDefaultBucket() {
     @Test
     void testWaitForFileToAppear(@TempDir final Path tempDir)
             throws BucketAccessException, InterruptedException, IOException, TimeoutException {
-        final String filename = "large-file.txt";
-        final Path tempFile = tempDir.resolve(filename);
+        final var filename = "large-file.txt";
+        final var tempFile = tempDir.resolve(filename);
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/test/java/com/exasol/bucketfs/BucketContentSyncIT.java:44]: Done.

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 7]
```diff
@@ -118,45 +123,82 @@ protected String extendPathInBucketDownToFilename(final Path localPath, final St
...
+            throw createDownloadIoException(uri, DOWNLOAD, exception);
+        } catch (final InterruptedException exception) {
+            Thread.currentThread().interrupt();
+            throw createDownloadInterruptedException(uri, DOWNLOAD);
         }
     }
 
-    private String httpGet(final URI uri) throws IOException, InterruptedException {
-        final HttpRequest request = HttpRequest.newBuilder(uri) //
+    private HttpRequest createRequest(final URI uri) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:146]: Maybe `createGetRequest` is better?

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java [hunk 8]
```diff
@@ -118,45 +123,82 @@ protected String extendPathInBucketDownToFilename(final Path localPath, final St
...
+    private BucketAccessException createDownloadInterruptedException(final URI uri, final BucketOperation operation) {
+        return new BucketAccessException(messageBuilder("E-BFSJ-4")
+                .message("Interrupted trying to {{operation|uq}} {{URI}}.", operation, uri).toString());
+    }
+
+    // [impl->dsn~downloading-a-file-from-a-bucket-as-string~1]
+    @Override
+    public String downloadFileAsString(final String pathInBucket) throws BucketAccessException {
+        final var uri = createPublicReadURI(pathInBucket);
+        LOGGER.fine(() -> "Downloading  file from bucket \"" + this + "\" at \"" + uri + "\"");
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:168]: I think it's better to use single quotes in log messages too to be consistent.

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
The build looks broken

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java (hunk 7 above)
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:146]: Fixed.

File: src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java (hunk 8 above)
Comment (redcatbear) [reply src/main/java/com/exasol/bucketfs/ReadEnabledBucket.java:168]: Done.

**Review by AnastasiiaSergienko — APPROVED** [review]

**Pull request merged**